src/main/scala/io/iohk/ethereum/network/rlpx/RLPxConnectionHandler.scala

```diff
@@ -74,39 +76,51 @@ class RLPxConnectionHandler(
-              // process pre-eip8 message
-              val remainingData = data.drop(InitiatePacketLength)
-              connection ! Write(responsePacket)
-              processHandshakeResult(result, remainingData)
+          val maybePreEIP8Result = Try {
+            val (responsePacket, result) = handshaker.handleInitialMessage(data.take(InitiatePacketLength))
+            val remainingData = data.drop(InitiatePacketLength)
+            (responsePacket, result, remainingData)
+          }
+          lazy val maybePostEIP8Result = Try {
+            val encryptedPayloadSize = ByteUtils.bigEndianToShort(data.take(2).toArray)
+            val (packetData, remainingData) = data.splitAt(encryptedPayloadSize + 2)
+            val (responsePacket, result) = handshaker.handleInitialMessageV4(packetData)
+            (responsePacket, result, remainingData)
+          }
 
-            case Failure(_) =>
-              // process as eip8 message
-              val encryptedPayloadSize = ByteUtils.bigEndianToShort(data.take(2).toArray)
-              val (packetData, remainingData) = data.splitAt(encryptedPayloadSize + 2)
-              val (responsePacket, result) = handshaker.handleInitialMessageV4(packetData)
+          maybePreEIP8Result orElse maybePostEIP8Result match {
+            case Success((responsePacket, result, remainingData)) =>
               connection ! Write(responsePacket)
               processHandshakeResult(result, remainingData)
+
+            case Failure(ex) =>
+              log.warning(s"[Stopping Connection] Init AuthHandshaker message handling failed for peer $peerId due to ${ex.getMessage}")
+              context.parent ! ConnectionFailed
+              context stop self
           }
       }
 
     def waitingForAuthHandshakeResponse(handshaker: AuthHandshaker, timeout: Cancellable): Receive =
       handleWriteFailed orElse handleTimeout orElse handleConnectionClosed orElse {
         case Received(data) =>
           timeout.cancel()
-          Try(handshaker.handleResponseMessage(data.take(ResponsePacketLength))) match {
-            case Success(result) =>
-              // process pre-eip8 message
-              val remainingData = data.drop(ResponsePacketLength)
-              processHandshakeResult(result, remainingData)
-
-            case Failure(_) =>
-              // process as eip8 message
-              val size = ByteUtils.bigEndianToShort(data.take(2).toArray)
-              val (packetData, remainingData) = data.splitAt(size + 2)
-              val result = handshaker.handleResponseMessageV4(packetData)
-              processHandshakeResult(result, remainingData)
+          val maybePreEIP8Result = Try {
+            val result = handshaker.handleResponseMessage(data.take(ResponsePacketLength))
+            val remainingData = data.drop(ResponsePacketLength)
+            (result, remainingData)
+          }
+          val maybePostEIP8Result = Try {
```

Comment:
this looks similar to line 84, is it possible to replace both with a method that returns Try?